x-pack/osquerybeat/ext/osquery-extension/pkg/amcache/registry/registry.go

```diff
@@ -0,0 +1,221 @@
+	if err != nil {
+		return nil, err
+	}
+	return fs.ReadFile(sourceFS, fsPath)
+}
+
+// loadExistingRegistry loads the registry from the given file path.  Without any transaction logs.
+func loadExistingRegistry(filePath string, log *logger.Logger) (*regparser.Registry, error) {
+	hiveContent, err := getFileContents(filePath, log)
+	if err != nil {
+		return nil, err
+	}
+	log.Infof("loaded registry from %s", filePath)
+	return regparser.NewRegistry(bytes.NewReader(hiveContent))
+}
+
+// createTempFile creates a temporary file with the given contents and returns the file handle.
+// caller is responsible for closing the file and removing the file after use. function returns
+// a handle that has been seeked to the start of the file.
+func createTempFile(contents []byte, log *logger.Logger) (*os.File, error) {
+	// create a temporary file
+	tempFile, err := os.CreateTemp("", "registry-*.hive")
+	if err != nil {
+		return nil, err
+	}
+
+	// write the contents to the temporary file
+	if _, err := tempFile.Write(contents); err != nil {
+		log.Errorf("failed to write to temp file: %s", err.Error())
+		return nil, err
+	}
+
+	// sync the temporary file
+	if err := tempFile.Sync(); err != nil {
+		log.Errorf("failed to sync temp file: %s", err.Error())
+		tempFile.Close()
+		os.Remove(tempFile.Name())
+		return nil, err
+	}
+
+	// seek to the start of the temporary file
+	_, err = tempFile.Seek(0, io.SeekStart)
+	if err != nil {
+		log.Errorf("failed to seek to start of temp file: %s", err.Error())
+		tempFile.Close()
+		os.Remove(tempFile.Name())
+		return nil, err
+	}
+	return tempFile, nil
+}
+
+// recoverRegistry recovers the registry from the given file path and transaction log paths.
+func recoverRegistry(filePath string, transactionLogPaths []string, log *logger.Logger) (registry *regparser.Registry, err error) {
+	// get the hive content
+	hiveContent, err := getFileContents(filePath, log)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a temporary file for the hive content
+	tempHiveFile, err := createTempFile(hiveContent, log)
+	if err != nil {
+		return nil, err
+	}
+	defer os.Remove(tempHiveFile.Name()) // Clean up temp file
+	defer tempHiveFile.Close()
+
+	// get the transaction log handles
+	transactionLogHandles := make([]*os.File, 0, len(transactionLogPaths))
+	for _, transactionLogPath := range transactionLogPaths {
+		transactionLogContent, err := getFileContents(transactionLogPath, log)
+		if err != nil {
+			log.Errorf("failed to get file contents for transaction log %s: %s", transactionLogPath, err.Error())
+			return nil, err
+		}
+		transactionLogHandle, err := createTempFile(transactionLogContent, log)
+		if err != nil {
+			log.Errorf("failed to create temporary file for transaction log %s: %s", transactionLogPath, err.Error())
+			return nil, err
+		}
+
+		// close and remove the temporary file
+		defer transactionLogHandle.Close()
+		defer os.Remove(transactionLogHandle.Name()) // Clean up temp file
```

Comment:
The order of these two statements looks backwards. You want the handle to be `Close()`d then `os.Removed`. And because `defer` is executed in reverse order these need flipped around.